Assets/Mirror/Runtime/Compression.cs

```diff
@@ -0,0 +1,258 @@
+    ///     <code>
+    ///     2/sqrt(2) / (2^bitCount - 1)
+    ///     </code>
+    /// </description></listheader>
+    /// 
+    /// <item><description>
+    ///     rotation precision +-0.00138 in range [-1,+1]
+    ///     <code>
+    ///     10 bits per value
+    ///     2 + 10 * 3 = 32 bits => send 4 bytes
+    ///     </code>
+    /// </description></item>
+    /// </list>
+    ///
+    /// <para>
+    /// Links for more info:
+    /// <br/><see href="https://youtu.be/Z9X4lysFr64">GDC Talk</see>
+    /// <br/><see href="https://gafferongames.com/post/snapshot_compression/">Post on Snapshot Compression</see>
+    /// </para>
+    /// </remarks>
+    public static class Compression
+    {
+        const float QuaternionMinValue = -1f / 1.414214f; // 1/ sqrt(2)
+        const float QuaternionMaxValue = 1f / 1.414214f;
+
+        const int QuaternionBitLength = 10;
+        // same as Mathf.Pow(2, targetBitLength) - 1
+        const uint QuaternionUintRange = (1 << QuaternionBitLength) - 1;
+
+        /// <summary>
+        /// Used to Compress Quaternion into 4 bytes
+        /// </summary>
+        public static uint CompressQuaternion(Quaternion value)
+        {
+            value = value.normalized;
+
+            int largestIndex = FindLargestIndex(value);
+            Vector3 small = GetSmallerDimensions(largestIndex, value);
+            // largest needs to be positive to be calculated by reader 
+            // if largest is negative flip sign of others because Q = -Q
+            if (value[largestIndex] < 0)
+            {
+                small *= -1;
+            }
+
+            uint a = ScaleToUInt(small.x, QuaternionMinValue, QuaternionMaxValue, 0, QuaternionUintRange);
+            uint b = ScaleToUInt(small.y, QuaternionMinValue, QuaternionMaxValue, 0, QuaternionUintRange);
+            uint c = ScaleToUInt(small.z, QuaternionMinValue, QuaternionMaxValue, 0, QuaternionUintRange);
+
+            // pack each 10 bits and extra 2 bits into uint32
+            uint packed = a | b << 10 | c << 20 | (uint)largestIndex << 30;
+
+            return packed;
+        }
+
+        internal static int FindLargestIndex(Quaternion q)
+        {
+            int index = default;
+            float current = default;
+
+            // check each value to see which one is largest (ignoring +-)
+            for (int i = 0; i < 4; i++)
+            {
+                float next = Mathf.Abs(q[i]);
+                if (next > current)
+                {
+                    index = i;
+                    current = next;
+                }
+            }
+
+            return index;
+        }
+
+        static Vector3 GetSmallerDimensions(int largestIndex, Quaternion value)
+        {
+            float x = value.x;
+            float y = value.y;
+            float z = value.z;
+            float w = value.w;
+
+            float a, b, c;
+            switch (largestIndex)
+            {
+                case 0:
+                    a = y;
```

Comment:
this function can be way shorter:

case 0: return new Vector3(y, z, w);

Reply:
The function was doing more after the switch, but not its not I will shorten this